nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -89,94 +119,361 @@ contract WorkLock {
         endBidDate = _endBidDate;
+        endCancellationDate = _endCancellationDate;
         boostingRefund = _boostingRefund;
         stakingPeriods = _stakingPeriods;
+        minAllowedBid = _minAllowedBid;
+        maxAllowableLockedTokens = escrow.maxAllowableLockedTokens();
+        minAllowableLockedTokens = escrow.minAllowableLockedTokens();
+        creator = msg.sender;
     }
 
     /**
     * @notice Deposit tokens to contract
     * @param _value Amount of tokens to transfer
     */
     function tokenDeposit(uint256 _value) external {
-        require(block.timestamp <= endBidDate, "Can't deposit more tokens after end of bidding");
+        require(block.timestamp < endBidDate, "Can't deposit more tokens after end of bidding");
         token.safeTransferFrom(msg.sender, address(this), _value);
         tokenSupply += _value;
         emit Deposited(msg.sender, _value);
+        require(tokenSupply / minAllowableLockedTokens <= uint128(0) - 1, "Potential overflow for bidder index");
     }
 
     /**
     * @notice Calculate amount of tokens that will be get for specified amount of ETH
     * @dev This value will be fixed only after end of bidding
     */
     function ethToTokens(uint256 _ethAmount) public view returns (uint256) {
-        return _ethAmount.mul(tokenSupply).div(ethSupply);
+        if (_ethAmount < minAllowedBid) {
+            return 0;
+        }
+
+        // when all participants bid with the same minimum amount of eth
+        if (bonusETHSupply == 0) {
+            return tokenSupply / bidders.length;
+        }
+
+        uint256 bonusETH = _ethAmount - minAllowedBid;
+        uint256 bonusTokenSupply = tokenSupply - bidders.length * minAllowableLockedTokens;
+        return minAllowableLockedTokens + bonusETH.mul(bonusTokenSupply).div(bonusETHSupply);
+    }
+
+    /**
+    * @notice Calculate amount of work that need to be done to refund specified amount of ETH
+    */
+    function ethToWork(uint256 _ethAmount, uint256 _tokenSupply, uint256 _ethSupply)
+        internal view returns (uint256)
+    {
+        return _ethAmount.mul(_tokenSupply).mul(SLOWING_REFUND).divCeil(_ethSupply.mul(boostingRefund));
+    }
+
+    /**
+    * @notice Calculate amount of work that need to be done to refund specified amount of ETH
+    * @dev This value will be fixed only after end of bidding
+    * @param _ethAmount Specified amount of ETH
+    * @param _remainderOfDepositedETH Base part of ETH amount for correct calculation of work
+    * @dev ethAmount + remainderOfDepositedETH = depositedETH
+    */
+    function ethToWork(uint256 _ethAmount, uint256 _remainderOfDepositedETH) internal view returns (uint256) {
+
+        uint256 baseETHSupply = bidders.length * minAllowedBid;
+        // when all participants bid with the same minimum amount of eth
+        if (bonusETHSupply == 0) {
+            return ethToWork(_ethAmount, tokenSupply, baseETHSupply);
+        }
+
+        uint256 baseETH = 0;
```

Comment:
```suggestion
        uint256 baseRateETH = 0;
```

Reply:
I'm not sure about this changes, because it's base part of bid and difficult to see here rate

Reply:
Yeh, I'm not thrilled about my suggestion, but can't think of anything better. It's important people understand that this variable equals the segment of the ETH deposit that can be reclaimed at a 'base' ETH/NU rate, not the rate itself.

Reply:
I'm not sure I see a "rate" either - we could make the variable more explanatory eg. `baseETHToReclaim` and `bonusETHToReclaim` since they are just respective proportions of the `_ethToReclaim`